File: handlers/auto_format.py
"""
Auto-format handler — PostToolUse handler for Edit/Write.

Runs ruff (Python) or biome (TS/JS) after file modifications.
Formatting must never block the agent — all errors are swallowed.
"""

from __future__ import annotations

import os
import subprocess
import sys

from .base import ALLOW, HookResult

RUFF_BIN = "/opt/homebrew/bin/ruff"
PY_EXTS = {".py"}
JS_EXTS = {".ts", ".tsx", ".js", ".jsx"}
SKIP_DIRS = {"node_modules", ".git", "__pycache__", ".venv", "venv", "dist", "build", ".next", ".worktrees"}


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    if tool_name not in ("Edit", "Write"):
        return ALLOW

    file_path = tool_input.get("file_path", "")
    if not file_path:
        return ALLOW

    file_path = os.path.realpath(file_path)
    if not os.path.isfile(file_path):
        return ALLOW
    if _in_skip_dir(file_path):
        return ALLOW
    if not _find_git_root(file_path):
        return ALLOW

    _, ext = os.path.splitext(file_path)
    ext = ext.lower()

    if ext in PY_EXTS:
        _format_python(file_path)
    elif ext in JS_EXTS:
        _format_js(file_path)

    return ALLOW


def _log(msg: str) -> None:
    print(f"[auto-format] {msg}", file=sys.stderr)


def _find_git_root(path: str) -> str | None:
    current = os.path.dirname(path) if os.path.isfile(path) else path
    while current != "/":
        if os.path.isdir(os.path.join(current, ".git")):
            return current
        current = os.path.dirname(current)
    return None


def _in_skip_dir(file_path: str) -> bool:
    return any(p in SKIP_DIRS for p in file_path.split(os.sep))


def _find_biome_config(file_path: str) -> str | None:
    current = os.path.dirname(file_path)
    while current != "/":
        for name in ("biome.json", "biome.jsonc"):
            if os.path.isfile(os.path.join(current, name)):
                return current
        current = os.path.dirname(current)
    return None


def _find_biome_bin(config_dir: str) -> str | None:
    local_bin = os.path.join(config_dir, "node_modules", ".bin", "biome")
    if os.path.isfile(local_bin):
        return local_bin
    try:
        subprocess.run(["biome", "--version"], capture_output=True, timeout=5)
        return "biome"
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return None


def _format_python(file_path: str) -> None:
    if not os.path.isfile(RUFF_BIN):
        return
    try:
        subprocess.run([RUFF_BIN, "check", "--fix", "--silent", file_path],
                       capture_output=True, text=True, timeout=15)
        result = subprocess.run([RUFF_BIN, "format", "--quiet", file_path],
                                capture_output=True, text=True, timeout=15)
        if result.returncode == 0:
            _log(f"ruff OK: {file_path}")
        else:
            _log(f"ruff error: {result.stderr.strip()}")
    except Exception as e:
        _log(f"ruff exception: {e}")


def _format_js(file_path: str) -> None:
    config_dir = _find_biome_config(file_path)
    if not config_dir:
        return
    biome_bin = _find_biome_bin(config_dir)
    if not biome_bin:
        return
    try:
        result = subprocess.run(
            [biome_bin, "check", "--write", "--no-errors-on-unmatched", file_path],
            capture_output=True, text=True, timeout=15, cwd=config_dir,
        )
        if result.returncode == 0:
            _log(f"biome OK: {file_path}")
        else:
            _log(f"biome error: {result.stderr.strip()}")
    except Exception as e:
        _log(f"biome exception: {e}")

File: internal/handlers/auto_format.go
// Package handlers — auto_format.go
// PostToolUse handler for Edit/Write events.
// Runs ruff (Python) or biome (TS/JS) after file modifications.
// Formatting must never block the agent — all errors are swallowed.
package handlers

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/core"
)

func init() {
	core.Register("PostToolUse", core.Entry{
		Matcher:    "Edit|Write",
		Handler:    autoFormatHandle,
		Critical:   false,
		ModuleName: "auto_format",
	})
}

var autoFormatSkipDirs = map[string]struct{}{
	"node_modules": {},
	".git":         {},
	"__pycache__":  {},
	".venv":        {},
	"venv":         {},
	"dist":         {},
	"build":        {},
	".next":        {},
	".worktrees":   {},
}

var autoFormatPyExts = map[string]struct{}{
	".py": {},
}

var autoFormatJsExts = map[string]struct{}{
	".ts":  {},
	".tsx": {},
	".js":  {},
	".jsx": {},
}

func autoFormatHandle(_, toolName string, toolInput map[string]any, _ string) core.HookResult {
	if toolName != "Edit" && toolName != "Write" {
		return core.Allow()
	}

	rawPath, _ := toolInput["file_path"].(string)
	if rawPath == "" {
		return core.Allow()
	}

	filePath, err := filepath.EvalSymlinks(rawPath)
	if err != nil {
		filePath = rawPath
	}
	filePath, err = filepath.Abs(filePath)
	if err != nil {
		return core.Allow()
	}

	info, err := os.Stat(filePath)
	if err != nil || info.IsDir() {
		return core.Allow()
	}

	if autoFormatInSkipDir(filePath) {
		return core.Allow()
	}

	if autoFormatFindGitRoot(filePath) == "" {
		return core.Allow()
	}

	ext := strings.ToLower(filepath.Ext(filePath))

	if _, ok := autoFormatPyExts[ext]; ok {
		autoFormatPython(filePath)
	} else if _, ok := autoFormatJsExts[ext]; ok {
		autoFormatJS(filePath)
	}

	return core.Allow()
}

func autoFormatInSkipDir(filePath string) bool {
	parts := strings.Split(filePath, string(os.PathSeparator))
	for _, p := range parts {
		if _, skip := autoFormatSkipDirs[p]; skip {
			return true
		}
	}
	return false
}

func autoFormatFindGitRoot(path string) string {
	current := filepath.Dir(path)
	for {
		if _, err := os.Stat(filepath.Join(current, ".git")); err == nil {
			return current
		}
		parent := filepath.Dir(current)
		if parent == current {
			return ""
		}
		current = parent
	}
}

func autoFormatFindBiomeConfig(filePath string) string {
	current := filepath.Dir(filePath)
	for {
		for _, name := range []string{"biome.json", "biome.jsonc"} {
			if _, err := os.Stat(filepath.Join(current, name)); err == nil {
				return current
			}
		}
		parent := filepath.Dir(current)
		if parent == current {
			return ""
		}
		current = parent
	}
}

func autoFormatFindBiomeBin(configDir string) string {
	localBin := filepath.Join(configDir, "node_modules", ".bin", "biome")
	if info, err := os.Stat(localBin); err == nil && !info.IsDir() {
		return localBin
	}
	p, err := exec.LookPath("biome")
	if err == nil {
		return p
	}
	return ""
}

func autoFormatPython(filePath string) {
	ruffBin := core.Cfg().GetTool("ruff")
	if ruffBin == "" {
		return
	}
	if _, err := os.Stat(ruffBin); err != nil {
		return
	}
	// ruff check --fix --silent
	core.RunCmd([]string{ruffBin, "check", "--fix", "--silent", filePath}, "", 15*time.Second, "")
	// ruff format --quiet
	core.RunCmd([]string{ruffBin, "format", "--quiet", filePath}, "", 15*time.Second, "")
}

func autoFormatJS(filePath string) {
	configDir := autoFormatFindBiomeConfig(filePath)
	if configDir == "" {
		return
	}
	biomeBin := autoFormatFindBiomeBin(configDir)
	if biomeBin == "" {
		return
	}
	core.RunCmd(
		[]string{biomeBin, "check", "--write", "--no-errors-on-unmatched", filePath},
		"", 15*time.Second, configDir,
	)
}
